### ISSUE TITLE
Low Raytracing performance on AMD

### ISSUE THREAD

**Issue by Try** [Try/OpenGothic#452]
Noted on a stream here: 
https://youtu.be/PMP53yaC7n0?t=9370

First idea is to make use of `VK_GEOMETRY_INSTANCE_FORCE_OPAQUE_BIT_KHR` and DX12 equivalent for opaque instances, yet exact reason why RX6600 is much-much slower than RTX3070 is unknown.

**Comment by Try** [comment]
Marking as help-wanted: need someone with RX6600(or other RT-capable AMD), to help with testing

**Comment by Nindaleth** [comment]
I have RX 6700 XT and could do the test. I don't really understand what's the issue in stream, though, is it about going indoors?

**Comment by Nindaleth** [comment]
Hi @Try :-) Actually yes, I've been getting the indoors slowdown for a very long time and just assumed everybody gets that so it's known :-D I never got around to measure it properly to make the issue detailed enough.

For a quick test of the patch, Standing in front of the Khorinis hotel I get around 72 FPS, going straight in until I stop at Hannah's desk I then get around 24 FPS. After applying the patch, unfortunately I don't see any difference in FPS nor in the visuals. Should something look much different? All Extended Configuration options are enabled, there are no special switches used when running the game.

**Comment by Try** [comment]
> Should something look much different?

Alpha-tested geometry should break, in RT-shaders. You can verify this by testing grass against torch.

> All Extended Configuration options are enabled

RT is not a menu option. Game simply enables it for discrete gpu's with RT-support, assuming that they are capable of simple shadows.
To test RT on/off you can use command line flags `-rt 0`

Also for, if you interested to have deep dive into shader: `shader/lighting/light.frag`, there is a function `bool isShadow(...)`, that implements simple ray-test loop.

Actually... `gl_RayFlagsNoOpaqueEXT` is wrong there, need to remove this as well, in order to make path usefull. Sorry found it, as been writing this post :D

**Comment by Nindaleth** [comment]
> RT is not a menu option. 

You're right, sorry, I mentioned the extended configuration in case the raytracing worked specially bad with some of the settings, performance-wise.

> Actually... `gl_RayFlagsNoOpaqueEXT` is wrong there, need to remove this as well, in order to make path usefull. Sorry found it, as been writing this post :D

Alright, I'm glad my lack of difference is expected :-D

Retested with the light.frag changed too. The torch+grass now looks bad so it works, but I still get the same 72 -> 24 FPS drop in Hanna's hotel test.

**Comment by Try** [comment]
> 72 -> 24 FPS drop

Eh, that's not good at all. 72 is RT+outdoor, right? What FPS in non-rt case?

PS
Decided to commit [e5a5fee](https://github.com/Try/OpenGothic/commit/e5a5fee42b7fe4b4ac274b218d31bccc2dbeb3ba) with proper RT-flags. At least now engine matches AMD guideline...

**Comment by Nindaleth** [comment]
Thanks, I've pulled the new changes. If you'd like me to test any patches on top of the current master, let me know what to try.

As for the testing, this is the outdoor part

![test_part1_rton](https://github.com/Try/OpenGothic/assets/6383928/76c7d317-bd65-41e0-9a3c-1e6d31d41324)

And this is the indoor part
![test_part2_rton](https://github.com/Try/OpenGothic/assets/6383928/a100a558-c140-46b4-8c1f-979be2364be4)

On non-RT the FPS is around the same and there's no change when moving indoors. Roughly:
* RT+outdoor = ~70 FPS
* RT+indoor = ~25 FPS
* -rt 0 + outdoor = ~70 FPS
* -rt 0 + indoor = ~70 FPS

**Comment by Nindaleth** [comment]
Did you consider running a GoFundMe or similar so that you're able to purchase your own RT-supporting Radeon and debug this? :D

**Comment by Nindaleth** [comment]
Oh OK, I see and agree.
I prefer a desktop so it would be a question of swapping just the GPU, no further purchases needed, but laptop form factor makes that part harder.

**Comment by Try** [comment]
Just for the record:
![изображение](https://github.com/Try/OpenGothic/assets/122894/18a4681c-a908-4f3c-a2d0-21b4dd869f66)

Polygon complexity is pretty-much non-existing, everything is `opaque`.
For some reason NSight show that none of acceleration-structures are `FAST_TRACE` (despite that the only mode engine actually supports), also none are FAST_BUILD. Maybe application bug, maybe NVidia bug here.

Amount of rays in scene is very high - tom of light-sources, on both floor of the building.

**Comment by Nindaleth** [comment]
Here's an interesting part on a [blogpost](https://pixelcluster.github.io/RADV-Raytracing-ON/) about the raytracing in RADV driver (what I use as a Radeon Vulkan driver on Linux):

> **How well do ray queries run?**
> Pretty competitive with AMDVLK/the AMD Windows drivers! You’ll generally see similar, if not better, performance on RADV.
>
> **How well do pipelines run?**
> Not well (expect significantly less performance compared to AMDVLK/Windows drivers). This is being worked on.

Are these RT pipelines something that's used OpenGothic? It would explain the less-than-expected performance.

**Comment by Try** [comment]
> Are these RT pipelines something that's used OpenGothic?

No, only rayQuery. There is a better resource explaining how they do thing: https://gpuopen.com/learn/improving-rt-perf-with-rra/

Overall, while exact algorithm that driver uses for traversal is not specified, probably it's sort-of tree with rope-nodes or similar.
What is quite bad for games, as such algorithm is set to work well on a few 3D models with high poly count, but not on large amount of low-poly.

In theory, it's possible to merge all static object into one giant mesh, to get some extra performance. Or we can wait for RTX-2.0 api, and see if it fixes anything

**Comment by Try** [comment]
Mac M1
<img width="1440" alt="Снимок экрана 2023-07-03 в 20 26 25" src="https://github.com/Try/OpenGothic/assets/122894/2acd043e-0750-4424-b3f2-efa30204f68e">

Well... could be worse :D

**Comment by Nindaleth** [comment]
> Well... could be worse :D

Could be 2 FPS less still :D What FPS do you get on your RTX in the hotel?

**Comment by Try** [comment]
Actually not that great: 41 fps. And RTCORE throughput is dominating frame:
![изображение](https://github.com/Try/OpenGothic/assets/122894/1d2ed9b3-2c5f-4c80-ad2c-5f93995ea838)

**Comment by Try** [comment]
Few ideas:
1. Remove some light-sources from RT, based on some criteria:
```
--- a/game/world/worldlight.cpp
+++ b/game/world/worldlight.cpp
@@ -4,7 +4,8 @@

 WorldLight::WorldLight(Vob* parent, World& world, const phoenix::vobs::light& vob, Flags flags)
   : Vob(parent,world,vob,flags) {
-  light = LightGroup::Light(world, vob);
+  if(vob.quality!=phoenix::light_quality::medium)
+    light = LightGroup::Light(world, vob);
   }

 void WorldLight::moveEvent() {
```
path above results in 60+ fps, unfortunately `quality` criteria is semi-random
2. Merge multiple `BLAS` into one big supper-`BLAS`.
cons: ugly; relies on how RT might be implemented on AMD/NVidia; no way to predict gains
3. Virtual shadow map
cons: one of the hardest raster-technique to implement; no MacOS

**Comment by Nindaleth** [comment]
Not sure what's the specific cause, but on the latest (74497771) I'm getting something like 70/30 FPS (instead of 72/24 previously).

I wanted to report getting even 45 FPS in the hotel indoors instead of 30 FPS, but it turns out the new resolution slider was set to 75% :D

**Comment by Nindaleth** [comment]
Current measurement of the same scene on latest cc35d6df (still the same hardware, still 1440p and 100% render resolution, all extended graphics features enabled): 80/35 FPS. For comparison, I get 105/57 FPS in 1080p. Not sure how it could improve so significantly, but I'm happy about the improvements.

**Comment by Try** [comment]
Hm, there were some optimization commits in recent time:
* 390490c0
* 8bbf61c1
* 8064c7c1
* de8fcb97
* 1173a3b8

Still bit too much to explain the difference... or maybe magic amd-driver update :D
